src/librustc_trans/trans/context.rs

```diff
@@ -977,11 +942,62 @@ fn declare_intrinsic(ccx: &CrateContext, key: &str) -> Option<ValueRef> {
                 ccx.intrinsics().borrow_mut().insert($name, f.clone());
                 return Some(f);
             }
+        );
+        ($name:expr, $cname:ident ($($arg:expr),*) -> $ret:expr) => (
+            if !is_like_pnacl {
+                // The `if key == $name` is already in ifn!
+                ifn!($name, fn($($arg),*) -> $ret);
+            } else if key == $name {
+                let f = declare::declare_cfn(ccx, stringify!($cname),
+                                             Type::func(&[$($arg),*], &$ret),
+                                             ccx.tcx().mk_nil());
+                ccx.intrinsics().borrow_mut().insert($name, f.clone());
+                return Some(f);
+            }
         )
     }
 
     compatible_ifn!("llvm.assume", noop(llvmcompat_assume(i1) -> void), 6);
 
+    // These must be redirected to `libm` here so PNaCl's linker will link in
+    // these functions from `libm`.
+    compatible_ifn!("llvm.copysign.f32", copysignf(t_f32, t_f32) -> t_f32);
```

Comment:
Sorry for the delay in replying to you.

> In other words, I believe that code sample with __builtin_powif is valid C code, and the LLVM IR derived from it is valid IR.

It isn't: functions starting with double underscores are implementation defined functions, according to the C/C++ standard.

Also, it won't just work: using `__builtin_powi` for example and compiling with `emcc` doesn't error, but still leaves an undefined function `llvm_powi_f32`, which matches what I found when I look through JSBackend. In order words, LLVM's intrinsics are not handled.

> If I understood what PNaCl does here, it first links all the bitcode, then it does lowering of intrinsics to the PNaCl final wire format. And some of those lowerings might require libm, which might not have been linked in before? It sounds like this is just a limitation in PNaCl - it should do something before that linking, to determine if it will need libm later, something like option 1 that you mention. PNaCl linking is just different enough from typical native linking that something like this is necessary, since natively, libm would be linked in after compiling to object/executable.

No, libm is not linked after translation. NaCl assumes `-mcpu=core2` for x86/64, so LLVM can lower

* llvm.bswap
* llvm.ctlz
* llvm.cttz
* llvm.ctpop
* llvm.fabs
* llvm.sqrt

directly into native instructions. Similarly for ARM and Mips. These intrinsics are the *only* math related intrinsics allow by the PNaCl ABI. All other math-y functions *must* be linked in via libm (or whatever) during the bitcode link. 

Those intrinsics are handled by `emcc`/`em++`, so there's no difference vs JS.

> (Emscripten btw has a similar issue of first linking all the bitcode, then doing compilation, and it also has a mechanism to handle such dependencies, that we need to be aware of early because they would otherwise only be discovered too late. We have them explicitly listed in a deps.json file. We actually don't have this issue with math functions, because of how we lower them, but we do have it with others.)

PNaCl can't do that because then such behaviour would be implementation dependant. Forcing linkage before hand is the better model anyway, because if JS is really desired, JS can be generated without any post codegen linking, barring some other voodoo needed to interop with normal JS.